Add asyncio.Executor matching concurrent.futures.Executor

jurca opened the pull request:
The Executor provides means for scheduling a executing asyncio tasks with capped parallelism. The asyncio Executor's API matches the concurrent.futures.Executor's API for consistency.

See also: https://discuss.python.org/t/add-task-pipeline-to-asyncio-with-capped-parallelism-and-lazy-input-reading/79292

I will add documentation once the API is agreed upon to prevent any differences between code and docs.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

jurca committed:
Add asyncio Executor

The Executor provides means for scheduling a executing asyncio tasks
with capped parallelism. The asyncio Executor's API matches the
concurrent.futures.Executor's API for consistency.

See also: https://discuss.python.org/t/add-task-pipeline-to-asyncio-with-capped-parallelism-and-lazy-input-reading/79292/2

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+
+class _WorkFunction[R](Protocol):
+    def __call__(self, *args, **kwargs) -> R | Awaitable[R]:
+        ...
```
```suggestion
class _WorkFunction[R, **P](Protocol):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R | Awaitable[R]:
        ...
```
If the module has type hints -- let's use it constantly everywhere

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+
+async def _run_work_item[R](work_item: _WorkItem[R]) -> R:
+    result = work_item.fn(*work_item.args, **work_item.kwargs)
+    if isawaitable(result):
```
I'm sorry, I disagree with accepting two function colors.
In my mind,  async executor should work with async functions only.
Processing sync functions without blocking calls inside don't make sense; they could be called without an executor.
Handling sync functions with blocking calls requires a `to_thread()` wrapper; it is not a subject of the proposed class.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+        fn: _WorkFunction[R],
+        *iterables: Iterable | AsyncIterable,
+        timeout: Optional[float] = None,
+        chunksize: int = 1,
```
This argument is never used; please drop it.
`asyncio.Executor` is not 100% compatible with `concurrent.futures.Executor` anyway.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+        tasks_in_flight_limit = len(self._workers) + self._input_queue.maxsize
+        resume_feeding = Event()
+
+        async def _feeder() -> None:
```
I think the function could be extracted into a method.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+                )
+                if remaining_time is not None and remaining_time <= 0:
+                    raise TimeoutError()
+                result = await wait_for(task, remaining_time)
```
Please replace `wait_for()` with `async with timeout()`.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+    while True:
+        work_item = await input_queue.get()
+        try:
+            if work_item is None:
```
How could it be None?

jurca replied:
`None` was used as a terminator, however, I see now that shutting the queue down works too. I've refactored it, thank you for pointing this out.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+                    item_future.set_exception(exception)
+                continue
+            if not item_future.cancelled():
+                item_future.set_result(result)
```
```suggestion
            else:
                if not item_future.cancelled():
                    item_future.set_result(result)
```

jurca replied:
After some tinkering, I managed to made the code more compact, but the new version cannot use try-except-else if it is to remain compact.

Please review the new version and consider which way you would prefer.

For easier comparison, here's a try-except-else version:

```python
    async def _worker(self) -> None:
        while True:
            try:
                work_item = await self._input_queue.get()
                item_future = work_item.future

                try:
                    if item_future.cancelled():
                        continue

                    task = create_task(work_item.fn(
                        *work_item.args,
                        **work_item.kwargs,
                    ))
                    await wait([task, item_future], return_when=FIRST_COMPLETED)
                    if item_future.cancelled():
                        task.cancel()
                        continue
                    result = task.result()
                except BaseException as exception:
                    if not item_future.cancelled():
                        item_future.set_exception(exception)
                else:
                    if not item_future.cancelled():
                        item_future.set_result(result)
                finally:
                    self._input_queue.task_done()
            except QueueShutDown:  # The executor has been shut down.
                break
```

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+    iterators = [aiter(async_iterable) for async_iterable in async_iterables]
+    while True:
+        try:
+            items = await gather(*[anext(iterator) for iterator in iterators])
```
```suggestion
            items =[anext(iterator) for iterator in iterators]
```
`await gather()` is not for free; it creates a task per item.
Let's use plain list comprehension here. 
If a user wants concurrency, he could create a task explicitly providing arguments.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+import time
+from inspect import isawaitable
+from typing import (Any, AsyncIterable, Awaitable, Iterable, NamedTuple,
+                    Optional, Protocol, cast)
```
`AsyncIterable`, `Iterable`, `Awaitable` live in `collections.abc` now, `Optional[T]` should be replaced with `T | None`.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+        ...
+
+
+class _WorkItem[R](NamedTuple):
```
Please use a dataclass instead of named tuple.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+        ]
+        self._feeders = set()
+
+    async def submit(
```
What do you think about context vars?
Now, workers work with a context implicitly copied during the `Executor` creation.
How could I submit a function with the context that I have at the moment of `submit()` call?

jurca replied:
While I have no strong opinions about context vars, I do see they can be useful.

I suggest adding a `context: Context | None = None` parameter to `submit()` and `map()` that will be set to `contextvars.copy_context()` if `None`, and propagating the context in `_WorkItem` to be used inside a worker when executing `fn`.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+                continue
+
+            try:
+                task = create_task(_run_work_item(work_item))
```
I don't know if the worker should create a task per processed item.
It looks like an overhead now.

jurca replied:
I understand, however, I think `create_task` is needed if I want to race `fn` against possible cancellation of the work item, and cancel `fn`'s execution if that happens.

If I were to remove `create_task` and just `await fn(...)`, it will prevent immediate cancellation of long-running work items, which could be undesirable for the users, e.g. if it would occur while leaving an Executor's context due to a raised exception.

Do you know of a way to achieve this behavior without using `create_task`?

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+
+        for _ in self._workers:
+            await self._input_queue.put(None)
+        self._input_queue.shutdown()
```
Could these three lines be reduced to bare `self._input_queue.shutdown()`?

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+        self._input_queue.shutdown()
+
+        if wait:
+            await gather(*self._workers)
```
Could a worker raise an exception? Should `.shutdown()` propagate the exception to a caller?

jurca replied:
Unless there is a bug in the implementation, a worker should never raise an exception.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+    return result
+
+
+async def _worker[R](input_queue: Queue[Optional[_WorkItem[R]]]) -> None:
```
Please make the function a private method of `Executor`.

asvetlov commented on Lib/asyncio/executor.py:
```diff
@@ -0,0 +1,226 @@
+                        _consume_cancelled_future(work_item.future),
+                    ))
+            if finalization_tasks:
+                await gather(*finalization_tasks)
```
If the code waits for all futures cancellations, there is no need to create a task per future.
`for` loop with waiting futures one by one could be enough; it doesn't wait the last future longer than the current code.

asvetlov reviewed (changes requested):
I've made a brief review concentrating on `await executor.submit(...)` method.
`map()` is another more complex beast; let's carefully work on it after basic `submit()` / `shutdown()` functionality is done.

jurca committed:
Add variadic type var for work function

jurca committed:
Require the WorkFunction to be async

Processing sync functions without blocking calls inside don't make
sense; they could be called without an executor.

Handling sync functions with blocking calls requires a to_thread()
wrapper; it is not a subject of the Executor class.

jurca committed:
Remove the unused chunksize parameter

The goal is not to match concurrent.futures.Executor's API 100%, and the
parameter does not really make sense in asyncio context.

jurca committed:
Refactor _feeder to a method

jurca committed:
Use async context manager to enforce task timeouts

jurca committed:
Use queue shutdown to terminate executor pipelines

As Queue.shutdown() was already used, sending a terminator None value is
essentially superfluous.

jurca committed:
Improve code style

jurca committed:
Improve map's input iterable combining performance

Using `gather` creates a task per item, introducing both performance and
memory overhead.

On the other hand, this removes the automagic concurrency if the calling
code provides multiple async iterables. This is deemed to be outside of
the Executor's responsibility and can be handled at the caller.

jurca committed:
Use up-to-date typing conventions

jurca committed:
Use a dataclass for work items

jurca committed:
Make _worker function an Executor's method

jurca committed:
Simplify task cancellation

jurca committed:
Remove no longer needed None checks
